src/serac/physics/solid_functional.hpp

```diff
@@ -215,108 +283,44 @@ class SolidFunctional : public BasePhysics {
-      mesh_.NewNodes(*reference_nodes_);
-    }
-
-    if (is_quasistatic_) {
-      // Update the time for housekeeping purposes
-      time_ += dt;
-      // Project the essential boundary coefficients
-      for (auto& bc : bcs_.essentials()) {
-        bc.setDofs(displacement_, time_);
-      }
-
-      quasiStaticSolve();
-    } else {
-      // Note that the ODE solver handles the essential boundary condition application itself
-      ode2_.Step(displacement_, velocity_, time_, dt);
-    }
-
-    if (geom_nonlin_ == GeometricNonlinearities::On) {
-      // Update the mesh with the new deformed nodes
-      deformed_nodes_->Set(1.0, displacement_.gridFunction());
-      deformed_nodes_->Add(1.0, *reference_nodes_);
-
-      mesh_.NewNodes(*deformed_nodes_);
-    }
-
-    cycle_ += 1;
-  }
-
   /**
    * @brief Set the material stress response and mass properties for the physics module
    *
    * @tparam MaterialType The solid material type
-   * @param material A material containing density and stress evaluation information
+   * @tparam StateType the type that contains the internal variables for MaterialType
+   * @param material A material that provides a function to evaluate stress
+   * @param qdata the buffer of material internal variables at each quadrature point
    *
-   * @pre MaterialType must have a method density() defining the density
-   * @pre MaterialType must have the operator (du_dX) defined as the Kirchoff stress
+   * @pre MaterialType must have a public member variable `density`
+   * @pre MaterialType must define operator() that returns the Kirchoff stress
    */
-  template <typename MaterialType>
-  void setMaterial(MaterialType material)
+  template <typename MaterialType, typename StateType>
+  void setMaterial(MaterialType material, std::shared_ptr<QuadratureData<StateType>> qdata)
   {
-    if constexpr (is_parameterized<MaterialType>::value) {
-      static_assert(material.numParameters() == sizeof...(parameter_space),
-                    "Number of parameters in solid does not equal the number of parameters in the "
-                    "solid material.");
-    }
-
-    auto parameterized_material = parameterizeMaterial(material);
-
-    K_functional_->AddDomainIntegral(
+    residual_->AddDomainIntegral(
         Dimension<dim>{},
-        [this, parameterized_material](auto x, auto displacement, auto... params) {
-          // Get the value and the gradient from the input tuple
-          auto [u, du_dX] = displacement;
-
-          auto source = zero{};
-
-          auto response = parameterized_material(x, u, du_dX, serac::get<0>(params)...);
-
-          auto flux = response.stress;
+        [this, material](auto /*x*/, auto& state, auto displacement, auto acceleration, auto... params) {
+          auto a          = get<VALUE>(acceleration);
+          auto du_dX      = get<DERIVATIVE>(displacement);
+          auto body_force = material.density * a + 0.0 * du_dX[0];
```

Comment:
This one still does something. The body force term here doesn't directly depend on the displacement, so it makes more sense to not have the ` + 0.0 * du_dX[0]` part initially. The problem is: if the geometric nonlinearity flag is turned on, then the body force is modified by multiplying by $\det(\boldsymbol{F})$ (making it dependent on the displacement). Since the geometric nonlinearity flag is a runtime choice, we can't write something like

```cpp
auto body_force = material.density * a;
auto stress     = material(state, du_dX, ...);
if (geom_nonlin_ == GeometricNonlinearities::On) {
  return {body_force * det(F), dot(stress, inv(transpose(F)))};
} else {
  return {body_force, stress};   
}
```
because the types of `body_force * det(F)` and `body_force` are different (and same with the stresses): the former tracks derivative information w.r.t. displacements, while the latter doesn't. C++ functions have to have a single return type, so you can't have a (runtime) conditional statement with branches that return different types.

Reply:
Per our discussion yesterday, the inertial force should not be modified as this is the reference density.